x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Space.java

```diff
@@ -0,0 +1,144 @@
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.ann.Fixed;
+import org.elasticsearch.compute.operator.BreakingBytesRefBuilder;
+import org.elasticsearch.compute.operator.EvalOperator.ExpressionEvaluator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.UnaryScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.common.unit.ByteSizeUnit.MB;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class Space extends UnaryScalarFunction {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Space", Space::new);
+
+    static final long MAX_LENGTH = MB.toBytes(1);
+
+    private final Expression number;
+
+    @FunctionInfo(
+        returnType = "keyword",
+        description = "Returns a string of `number` spaces.",
+        examples = @Example(file = "string", tag = "space")
+    )
+    public Space(
+        Source source,
+        @Param(name = "number", type = { "integer" }, description = "Number of spaces in result.") Expression number
+    ) {
+        super(source, number);
+        this.number = number;
+    }
+
+    private Space(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(number);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isType(number, dt -> dt == DataType.INTEGER, sourceText(), DEFAULT, "integer");
+    }
+
+    @Override
+    public boolean foldable() {
+        return number.foldable();
+    }
+
+    @Evaluator(extraName = "Constant")
+    static BytesRef processConstant(@Fixed(includeInToString = false, build = true) BreakingBytesRefBuilder scratch, @Fixed int number) {
+        return processInner(scratch, number);
+    }
```

Comment:
I'm not sure the `Constant` customization buys you much in this case. If you were to build an array of spaces up front and just append them then it'd help some. Not sure if it's worth it, but it doesn't hurt to have it.

Reply:
One place where it *can* really help is for primitive results when the "constant" flavor is infallible and the non-constant flavor is fallible. In that case we have to handle nullable results to deal with the warnings only in the one arm. That's less imporant for `BytesRef` to be honest, but maybe it's a thing? I'm not sure without a microbenchmark and am pretty sure it's not worth it for `SPACE`.

Reply:
Okay. Based on`Repeat#toEvaluator`, I did think the benefit was to validate parameter expressions early and fail hard when possible, rather than generate warnings and nulls later.

For `SPACE`, when `number` is foldable all the parameter validation can all be done in `toEvaluator`, but I guess it's not infallible because the circuit breaker can still break?

It sounds like performance benefits likely aren't worth it, but shall I remove this and have more warnings/nulls instead of failures?

Reply:
> Okay. Based on`Repeat#toEvaluator`, I did think the benefit was to validate parameter expressions early and fail hard when possible, rather than generate warnings and nulls later.

Actually, that's pretty nice. I hadn't thought about that. I think you could get the same check with, like `if (count.foldable()) {
checkTheNumber}`. But it's probably fine either way.

I guess I'd probably just push it all the way and build the spaces one time as a constant if the parameter is a constant. If you did that, could you do something like `return toEvaluator.apply(new Literal(source(), spaces, KEYWORD))`. That could work, right? We already know how to evaluate literals and there's nothing left to do with SPACES(constant) at runtime.

> but I guess it's not infallible because the circuit breaker can still break?

Those are still infallible because the entire request fails in that case. It only fallible in the collective "oh no, it's dead" way which doesn't really count.